README.md
# AI

Helpers for using AI in Elixir. Currently requires the [OpenAI](https://github.com/mgallo/openai.ex) package.

Includes `~l` sigil and `chat/1` function for OpenAI generation.

```elixir
iex> ~l"model:gpt-3.5-turbo user: how do I build an igloo in 10 words?"
[
  model: "gpt-3.5-turbo",
  messages: [%{role: "user", content: "how do I build an igloo in 10 words?"}]
]
```

```elixir
iex> ~l"model:gpt-3.5-turbo user: how do I build an igloo in 10 words?" |> chat()
{:ok, "Stack snow blocks in circle, decreasing size upward, till enclosed"}
```

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed by adding `ai` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ai, "~> 0.3.4"},
  ]
end
```

## Why?

[Here's why!](https://www.charlieholtz.com/articles/elixir-ai)


CHANGELOG.md
## [0.3.4] - 2022-01-01

### Changed

- Recommend using `~l` instead of `~LLM`, as it allows for interpolation.


test/ai_test.exs
defmodule AITest do
  use ExUnit.Case
  doctest AI
  import AI

  test "basic usage" do
    assert ~l"model:gpt-3.5-turbo user:how do I build an igloo in 10 words?" ==
             [
               model: "gpt-3.5-turbo",
               messages: [%{role: "user", content: "how do I build an igloo in 10 words?"}]
             ]
  end

  test "code interpolation works" do
    model = "gpt-3.5-turbo"
    assert ~l"model:#{model} user:how do I build an igloo in 10 words?" ==
             [
               model: "gpt-3.5-turbo",
               messages: [%{role: "user", content: "how do I build an igloo in 10 words?"}]
             ]
  end


end


test/potions_test.exs
defmodule PotionsTest do
  use ExUnit.Case
  doctest Potions
  import Potions

  test "greets the world" do
    assert ~LLM"model:gpt-3.5-turbo user:how do I build an igloo in 10 words?" ==
             [
               model: "gpt-3.5-turbo",
               messages: [%{role: "user", content: "how do I build an igloo in 10 words?"}]
             ]
  end
end


mix.exs
defmodule AI.MixProject do
  use Mix.Project

  def project do
    [
      app: :ai,
      version: "0.3.4",
      elixir: "~> 1.15",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      name: "AI",
      source_url: "https://github.com/cbh123/elixir_ai",
      homepage_url: "https://hexdocs.pm/ai/readme.html",
      package: package(),
      description: description(),
      docs: [
        main: "readme",
        extras: ["README.md", "CHANGELOG.md"]
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:openai, "~> 0.5.2"},
      {:ex_doc, "~> 0.14", only: :dev, runtime: false}
    ]
  end

  defp description() do
    "Helper functions for using AI in Elixir"
  end

  defp package() do
    [
      name: "ai",
      files: ~w(lib .formatter.exs mix.exs README*),
      licenses: ["Apache-2.0"],
      links: %{"GitHub" => "https://github.com/cbh123/elixir_ai"}
    ]
  end
end


lib/ai.ex
defmodule AI do
  @moduledoc """
  Documentation for `AI`.
  """

  @doc """
  Implements the ~l sigil, which parses text into an OpenAI friendly chat completion prompt.
  ~l works by parsing out the `model`, `system`, `user` and `assistant` keywords.

  Supports string interpolation.

  ## Examples
  iex> import AI
  iex> ~l"model: gpt-3.5-turbo system: You are an expert at text to image prompts. Given a description, write a text-to-image prompt. user: sunset"
  [
    model: "gpt-3.5-turbo",
    messages: [
      %{
        content: "You are an expert at text to image prompts. Given a description, write a text-to-image prompt.",
        role: "system"
      },
      %{content: "sunset", role: "user"}
    ]
  ]
  """
  def sigil_l(lines, _opts) do
   lines |> text_to_prompts()
  end

  @doc """
  DEPRECATED: Use ~l instead. ~LLM doesn't work with string interpolation.

  Implements the ~LLM sigil, which parses text into an OpenAI friendly chat completion prompt.
  ~LLM works by parsing out the `model`, `system`, `user` and `assistant` keywords.

  ## Examples
  iex> import AI
  iex> ~LLM"model: gpt-3.5-turbo system: You are an expert at text to image prompts. Given a description, write a text-to-image prompt. user: sunset"
  [
    model: "gpt-3.5-turbo",
    messages: [
      %{
        content: "You are an expert at text to image prompts. Given a description, write a text-to-image prompt.",
        role: "system"
      },
      %{content: "sunset", role: "user"}
    ]
  ]
  """

  @keywords "model:|system:|user:|assistant:"

  def sigil_LLM(lines, _opts) do
    lines |> text_to_prompts()
  end

  defp text_to_prompts(text) when is_binary(text) do
    extract_prompts(text)
    |> model_messages()
  end

  defp model_messages([%{content: model_name} | messages]) do
    [model: model_name, messages: messages]
  end

  defp extract_prompts(text) do
    pattern = ~r/(?<key>#{@keywords})\s*(?<value>(?:(?!#{@keywords}).)+)/

    Regex.scan(pattern, text)
    |> Enum.map(&convert_to_map/1)
  end

  defp convert_to_map([_, key, value]) do
    %{role: String.trim(key, ":"), content: String.trim(value)}
  end

  @doc """
  Parses out OpenAI's chat completion response into a cleaner format.
  Returns {:ok, text_content} or {:error, message}

  Instead of:
  {:ok,
  %{
   id: "chatcmpl-7zSc1rsCXpyALMjM9MkaF077xYRot",
   usage: %{
     "completion_tokens" => 10,
     "prompt_tokens" => 19,
     "total_tokens" => 29
   },
   created: 1694882349,
   choices: [
     %{
       "finish_reason" => "stop",
       "index" => 0,
       "message" => %{
         "content" => "Compact and stack snow blocks in a dome shape.",
         "role" => "assistant"
       }
     }
   ],
   model: "gpt-3.5-turbo-0613",
   object: "chat.completion"
  }}
  """
  def chat(text) do
    text
    |> OpenAI.chat_completion()
    |> parse_chat()
  end

  defp parse_chat({:ok, %{choices: [%{"message" => %{"content" => content}} | _]}}),
    do: {:ok, content}

  defp parse_chat({:error, %{"error" => %{"message" => message}}}), do: {:error, message}
end


lib/potions.ex
defmodule Potions do
  @moduledoc """
  Documentation for `Potions`.
  """

  @doc """
  Implements the ~LLM sigil, which parses text into an OpenAI friendly chat completion prompt.
  ~LLM works by parsing out the `model`, `system`, `user` and `assistant` keywords.

  ## Examples
  iex> import Potions
  iex> ~LLM"model: gpt-3.5-turbo system: You are an expert at text to image prompts. Given a description, write a text-to-image prompt. user: sunset"
  [
    model: "gpt-3.5-turbo",
    messages: [
      %{
        content: "You are an expert at text to image prompts. Given a description, write a text-to-image prompt.",
        role: "system"
      },
      %{content: "sunset", role: "user"}
    ]
  ]
  """
  def sigil_LLM(lines, _opts) do
    lines |> text_to_prompts()
  end

  defp text_to_prompts(text) when is_binary(text) do
    model = extract_model(text) |> String.trim()
    messages = extract_messages(text)
    [model: model, messages: messages]
  end

  defp extract_model(text) do
    extract_value_after_keyword(text, "model:")
  end

  defp extract_messages(text) do
    keywords = ["system:", "user:", "assistant:"]

    Enum.reduce_while(keywords, [], fn keyword, acc ->
      case extract_value_after_keyword(text, keyword) do
        nil ->
          {:cont, acc}

        value ->
          role = String.trim(keyword, ":")
          acc = acc ++ [%{role: role, content: String.trim(value)}]
          {:cont, acc}
      end
    end)
  end

  defp extract_value_after_keyword(text, keyword) do
    pattern = ~r/#{keyword}\s*(.*?)(?=model:|system:|user:|assistant:|$)/s

    case Regex.run(pattern, text) do
      [_, value] -> value
      _ -> nil
    end
  end

  @doc """
  Parses out OpenAI's chat completion response into a cleaner format.
  Returns {:ok, text_content} or {:error, message}

  Instead of:
  {:ok,
  %{
   id: "chatcmpl-7zSc1rsCXpyALMjM9MkaF077xYRot",
   usage: %{
     "completion_tokens" => 10,
     "prompt_tokens" => 19,
     "total_tokens" => 29
   },
   created: 1694882349,
   choices: [
     %{
       "finish_reason" => "stop",
       "index" => 0,
       "message" => %{
         "content" => "Compact and stack snow blocks in a dome shape.",
         "role" => "assistant"
       }
     }
   ],
   model: "gpt-3.5-turbo-0613",
   object: "chat.completion"
  }}
  """
  def chat(text) do
    text
    |> OpenAI.chat_completion()
    |> parse_chat()
  end

  defp parse_chat({:ok, %{choices: [%{"message" => %{"content" => content}} | _]}}),
    do: {:ok, content}

  defp parse_chat({:error, %{"error" => %{"message" => message}}}), do: {:error, message}
end
